metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,199 @@
+                     tax_tree_filepath, num_taxid_failed, relevant_taxids.size());
+    }
+
+    // Construct the output tree.
+    for (const TaxId &taxid : relevant_taxids) {
+        if (taxid == this->root_node) {
+            continue;
+        }
+        (*tree)[full_parents_list[taxid]].push_back(taxid);
+    }
+}
+
+void TaxonomyClsAnno::dfs_statistics(const TaxId node,
+                                     const ChildrenList &tree,
+                                     std::vector<TaxId> *tree_linearization) {
+    this->node_to_linearization_idx[node] = tree_linearization->size();
+    tree_linearization->push_back(node);
+    uint32_t depth = 0;
+    for (const TaxId &child : tree.at(node)) {
+        dfs_statistics(child, tree, tree_linearization);
+        tree_linearization->push_back(node);
+        if (this->node_depth[child] > depth) {
+            depth = this->node_depth[child];
+        }
+    }
+    this->node_depth[node] = depth + 1;
+}
+
+void TaxonomyClsAnno::rmq_preprocessing(const std::vector<TaxId> &tree_linearization) {
+    uint32_t num_rmq_rows = log2(tree_linearization.size()) + 1;
+
+    this->rmq_data.resize(num_rmq_rows);
+    for (uint32_t i = 0; i < num_rmq_rows; ++i) {
+        this->rmq_data[i].resize(tree_linearization.size());
+    }
+
+    // Copy tree_linearization to rmq[0].
+    for (uint32_t i = 0; i < tree_linearization.size(); ++i) {
+        this->rmq_data[0][i] = tree_linearization[i];
+    }
+
+    // Delta represents the size of the RMQ's sliding window (always a power of 2).
+    uint32_t delta = 1;
+    for (uint32_t row = 1; row < num_rmq_rows; ++row) {
+        for (uint32_t i = 0; i + delta < tree_linearization.size(); ++i) {
+            // rmq_data[row][i] covers an interval of size delta=2^row and returns the node with the maximal depth among positions [i, i+2^row-1] in the linearization.
+            // According to 'this->dfs_statistics()': node_depth[leaf] = 1 and node_depth[root] = maximum distance to a leaf.
+            if (this->node_depth[this->rmq_data[row - 1][i]] >
+                this->node_depth[this->rmq_data[row - 1][i + delta]]) {
+                this->rmq_data[row][i] = this->rmq_data[row - 1][i];
+            } else {
+                this->rmq_data[row][i] = this->rmq_data[row - 1][i + delta];
+            }
+        }
+        delta *= 2;
+    }
+
+    // Compute fast tables for log2 and pow2.
+    this->fast_log2.resize(tree_linearization.size());
+    this->fast_pow2.push_back(1);
+    for (uint32_t i = 2; i < tree_linearization.size(); ++i) {
+        this->fast_log2[i] = 1 + this->fast_log2[i/2];
+        if (this->fast_log2[i] > this->fast_log2[i-1]) {
+            this->fast_pow2.push_back(i);
+        }
+    }
+}
+
+TaxId TaxonomyClsAnno::assign_class(const std::string &sequence) const {
+    std::cerr << "assign class not implemented " << sequence << "\n\n";
+    return 0;
```

Comment:
This'll be hard to catch if people are redirecting stderr to some other place.

Use `throw std::runtime_error` instead